lib/model/message.dart

```diff
@@ -1,21 +1,332 @@
+  /// [kSendMessageOfferRestoreWaitPeriod] since the outbox message's creation.
+  waitPeriodExpired,
+
+  /// The message could not be delivered.
+  ///
+  /// This state can be reached when we got a 4xx or other error in the HTTP
+  /// response.
+  failed,
+}
+
+/// A message sent by the self-user.
+sealed class OutboxMessage<T extends Conversation> extends MessageBase<T> {
+  OutboxMessage({
+    required this.localMessageId,
+    required int selfUserId,
+    required super.timestamp,
+    required this.content,
+  }) : _state = OutboxMessageState.hidden,
+       super(senderId: selfUserId);
+
+  /// As in [MessageEvent.localMessageId].
+  ///
+  /// This uniquely identifies this outbox message's corresponding message object
+  /// in events from the same event queue.
+  ///
+  /// See also:
+  ///  * [MessageStoreImpl.sendMessage], where this ID is assigned.
+  final int localMessageId;
+  @override
+  int? get id => null;
+  final String content;
+
+  OutboxMessageState get state => _state;
+  OutboxMessageState _state;
+
+  /// Whether the [OutboxMessage] is hidden to [MessageListView] or not.
+  bool get hidden => _state == OutboxMessageState.hidden;
+}
+
+class StreamOutboxMessage extends OutboxMessage<StreamConversation> {
+  StreamOutboxMessage({
+    required super.localMessageId,
+    required super.selfUserId,
+    required super.timestamp,
+    required this.conversation,
+    required super.content,
+  });
+
+  @override
+  final StreamConversation conversation;
+}
+
+class DmOutboxMessage extends OutboxMessage<DmConversation> {
+  DmOutboxMessage({
+    required super.localMessageId,
+    required super.selfUserId,
+    required super.timestamp,
+    required this.conversation,
+    required super.content,
+  }) : assert(conversation.allRecipientIds.contains(selfUserId));
+
+  @override
+  final DmConversation conversation;
+}
+
+/// Manages the outbox messages portion of [MessageStore].
+mixin _OutboxMessageStore on PerAccountStoreBase {
+  late final UnmodifiableMapView<int, OutboxMessage> outboxMessages =
+    UnmodifiableMapView(_outboxMessages);
+  final Map<int, OutboxMessage> _outboxMessages = {};
+
+  /// A map of timers to show outbox messages after a delay,
+  /// indexed by [OutboxMessage.localMessageId].
+  ///
+  /// If the send message request fails within the time limit,
+  /// the outbox message's timer gets removed and cancelled.
+  final Map<int, Timer> _outboxMessageDebounceTimers = {};
+
+  /// A map of timers to update outbox messages state to
+  /// [OutboxMessageState.waitPeriodExpired] after a delay,
+  /// indexed by [OutboxMessage.localMessageId].
+  ///
+  /// If the send message request fails within the time limit,
+  /// the outbox message's timer gets removed and cancelled.
+  final Map<int, Timer> _outboxMessageWaitPeriodTimers = {};
+
+  /// A fresh ID to use for [OutboxMessage.localMessageId],
+  /// unique within this instance.
+  int _nextLocalMessageId = 0;
```

Comment:
Let's start at 1 rather than 0. That matches the practice in all the server's autoincrementing database indexes.